java/ql/src/experimental/Security/CWE/CWE-073/PathSanitizer.qll

```diff
@@ -0,0 +1,197 @@
+    e = super.getCheckedExpr() and
+    branch = false and
+    exists(BlockListGuard previousGuard |
+      DataFlow::localExprFlow(previousGuard.getCheckedExpr(), e) and
+      previousGuard.controls(this.getBasicBlock(), false)
+    )
+  }
+}
+
+/**
+ * Holds if `ma` is a call to a method that checks a partial string match.
+ */
+private predicate isStringPartialMatch(MethodAccess ma) {
+  ma.getMethod().getDeclaringType() instanceof TypeString and
+  ma.getMethod().getName() =
+    ["contains", "startsWith", "matches", "regionMatches", "indexOf", "lastIndexOf"]
+}
+
+/**
+ * Holds if `ma` is a call to a method of `java.nio.file.Path` that checks a partial path match.
+ */
+private predicate isPathPartialMatch(MethodAccess ma) {
+  ma.getMethod().getDeclaringType() instanceof TypePath and
+  ma.getMethod().getName() = "startsWith"
+}
+
+private predicate isDisallowedWord(CompileTimeConstantExpr word) {
+  word.getStringValue().matches(["%WEB-INF%", "%META-INF%", "%..%"])
+}
+
+/** A complementary guard that protects against path traversal, by looking for the literal `..`. */
+class PathTraversalGuard extends Guard instanceof MethodAccess {
+  Expr checked;
+
+  PathTraversalGuard() {
+    super.getMethod().getDeclaringType() instanceof TypeString and
+    super.getMethod().hasName(["contains", "indexOf"]) and
+    super.getAnArgument().(CompileTimeConstantExpr).getStringValue() = ".."
+  }
+
+  Expr getCheckedExpr() { result = super.getQualifier() }
+}
+
+/** A complementary sanitizer that protects against path traversal using path normalization. */
+private class PathNormalizeSanitizer extends MethodAccess {
+  PathNormalizeSanitizer() {
+    this.getMethod().getDeclaringType().hasQualifiedName("java.nio.file", "Path") and
+    this.getMethod().hasName("normalize")
+  }
+}
+
+/** A complementary guard that protects against double URL encoding, by looking for the literal `%`. */
+private class UrlEncodingGuard extends Guard instanceof MethodAccess {
+  UrlEncodingGuard() {
+    super.getMethod().getDeclaringType() instanceof TypeString and
+    super.getMethod().hasName(["contains", "indexOf"]) and
+    super.getAnArgument().(CompileTimeConstantExpr).getStringValue() = "%"
+  }
+
+  Expr getCheckedExpr() { result = super.getQualifier() }
+}
+
+/** A complementary sanitizer that protects against double URL encoding using URL decoding. */
+private class UrlDecodeSanitizer extends MethodAccess {
+  UrlDecodeSanitizer() {
+    this.getMethod().getDeclaringType().hasQualifiedName("java.net", "URLDecoder") and
+    this.getMethod().hasName("decode")
+  }
+}
+
+/** A sanitized node that is protected against path traversal vulnerabilities. */
+abstract class SanitizedNode extends DataFlow::Node { }
+
+class NodeWithPathNormalizer extends SanitizedNode {
+  NodeWithPathNormalizer() {
+    exists(MethodAccess ma |
+      DataFlow::localExprFlow(this.asExpr(), ma) and ma instanceof PathNormalizeSanitizer
```

Comment:
```suggestion
    DataFlow::localExprFlow(this.asExpr(), any(PathNormalizeSanitizer ma))
```

Reply:
Yes, it works with my test. Without the check of `and not sink instanceof SanitizedNode`, the test case `uploadFile2` will be wrongly included in query results as a bad case.

Reply:
And I've made the change to simplify the local expression flow.

Reply:
Ok, now I see why it works. You have this in `PathSanitizer.qll`:

```codeql
private class PathDataModel extends SummaryModelCsv {
  override predicate row(string row) {
    row =
      [
        "java.nio.file;Paths;true;get;;;Argument[0];ReturnValue;value",
        "java.nio.file;Path;true;normalize;;;Argument[-1];ReturnValue;value"
      ]
  }
}
```

Which isn't actually correct, because neither of those methods are actually value-preserving steps (there's a type change in the first case, and a value change in the second one). These should be taint steps, and thus you should use `TaintTracking::localExprTaint` in `SanitizedNode` if you wanted to keep track of them. By the way, `SanitizedNode` has a rather misleading name, because it's not a sanitizer, and in fact is in some way duplicating the logic of `AllowListBarrierGuard` + `PathNormalizeSanitizer`. 

The whole point is that the sink shouldn't be `Paths.get` but rather `new FileOutputStream(path.toFile())`, which comes after the check and therefore is protected by `AllowListBarrierGuard`. But this is something that we can look into during an eventual promotion + mashup of this query with `java/path-injection`.

Reply:
I've renamed <code>SanitizedNode</code> to <code>NormalizedPathNode</code> since that is what it is and changed those methods to taint steps. I agree the sink shall not be <code>Paths.get</code>, which will address the issue elegantly, and will leave it to the expert when doing the eventual promotion + mashup of this query with <code>java/path-injection</code>:-)